homeassistant/components/denonavr/config_flow.py

```diff
@@ -0,0 +1,234 @@
+        vol.Optional(CONF_TIMEOUT, default=DEFAULT_TIMEOUT): vol.All(
+            int, vol.Range(min=1)
+        ),
+        vol.Optional(CONF_SHOW_ALL_SOURCES, default=DEFAULT_SHOW_SOURCES): bool,
+        vol.Optional(CONF_ZONE2, default=DEFAULT_ZONE2): bool,
+        vol.Optional(CONF_ZONE3, default=DEFAULT_ZONE3): bool,
+    }
+)
+
+
+class DenonAvrFlowHandler(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a Denon AVR config flow."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_POLL
+
+    def __init__(self):
+        """Initialize the Denon AVR flow."""
+        self.host = None
+        self.serial_number = None
+        self.model_name = None
+        self.timeout = DEFAULT_TIMEOUT
+        self.show_all_sources = DEFAULT_SHOW_SOURCES
+        self.zone2 = DEFAULT_ZONE2
+        self.zone3 = DEFAULT_ZONE3
+        self.d_receivers = []
+
+    async def async_step_user(self, user_input=None):
+        """Handle a flow initialized by the user."""
+        errors = {}
+        if user_input is not None:
+            # check if IP address is set manually
+            host = user_input.get(CONF_HOST)
+            if host:
+                self.host = host
+                return await self.async_step_settings()
+
+            # discovery using denonavr library
+            self.d_receivers = await self.hass.async_add_executor_job(denonavr.discover)
+            # More than one receiver could be discovered by that method
+            if len(self.d_receivers) == 1:
+                self.host = self.d_receivers[0]["host"]
+                return await self.async_step_settings()
+            if len(self.d_receivers) > 1:
+                # show selection form
+                return await self.async_step_select()
+
+            errors["base"] = "discovery_error"
+
+        return self.async_show_form(
+            step_id="user", data_schema=CONFIG_SCHEMA, errors=errors
+        )
+
+    async def async_step_select(self, user_input=None):
+        """Handle multiple receivers found."""
+        errors = {}
+        if user_input is not None:
+            self.host = user_input["select_host"]
+            return await self.async_step_settings()
+
+        select_scheme = vol.Schema(
+            {
+                vol.Required("select_host"): vol.In(
+                    [d_receiver["host"] for d_receiver in self.d_receivers]
+                )
+            }
+        )
+
+        return self.async_show_form(
+            step_id="select", data_schema=select_scheme, errors=errors
+        )
+
+    async def async_step_settings(self, user_input=None):
+        """Allow the user to specify settings."""
+        if user_input is not None:
+            # Get config option that have defaults
+            self.timeout = user_input[CONF_TIMEOUT]
+            self.show_all_sources = user_input[CONF_SHOW_ALL_SOURCES]
+            self.zone2 = user_input[CONF_ZONE2]
+            self.zone3 = user_input[CONF_ZONE3]
```

Comment:
These should likely be moved to an options flow in a followup PR.

Reply:
I was not aware of the existance of the options flow when I made this PR.
Will probably do that in the future, but first want this PR to be done.

Reply:
The config flow removes support for multiple zones in the current state. Several Denon receivers support multiple zones with different speakers set to different volumens and listening to different sources. In fact separate media player instances.
As far as I understand the concept of devices and entities, there should be 1 device for the Denon receiver with up to 3 media player entities for each enabled zone.

Reply:
@scarface-4711 that is also the case, you still get 3 entities for the 3 diffrent zones if you check the zone2 and zone3 boxes

Reply:
So it does not remove support for multiple zones, those are still supported

Reply:
I think you are right right 😄 
The debugger shows how the additional zones are queried. They probably do not appear because of the MAC adress issue, I just posted.